File: Astro-TFG/public/scripts/4-manejo-peticiones/prueba-1/version-1/javascript/main.js
let worker = null;
let workerTime = 0;

function initializeWorker() {
    if (!worker) {
        worker = new Worker(
            window.location.origin +
            window.document.body.dataset.jsPath +
            "worker.js"
        );
    }
}

async function _runJsBenchmark() {
    try {
        window.clearCell("javascript-output");
        window.showExecutionLoader();

        const startWorkerTime = performance.now();
        await initializeWorker();
        workerTime = performance.now() - startWorkerTime;

        const num_requests = parseInt(
            document.getElementById("num-requests-javascript").value,
            10
        );
        const delay = parseInt(
            document.getElementById("request-delay-javascript").value,
            10
        );

        const id = `js-${Date.now()}`;
        const resultJson = await new Promise((resolve, reject) => {
            function onMessage(e) {
                if (e.data.id !== id) return;
                worker.removeEventListener("message", onMessage);
                if (e.data.error) {
                    reject(new Error(e.data.error));
                } else if (e.data.json !== undefined) {
                    resolve(e.data.json);
                } else {
                    reject(new Error("Worker dont return results"));
                }
            }
            worker.addEventListener("message", onMessage);
            worker.postMessage({
                id,
                type: "do_analisis",
                num_requests,
                delay,
            });
        });

        const result = JSON.parse(resultJson);
        displayResult(result, num_requests);
    } catch (e) {
        console.error(e);
    } finally {
        window.hideExecutionLoader();
    }
}

function createDiv() {
    const div = document.createElement("div");
    return div;
}

function displayResult(r, total_requests) {
    const output = document.getElementById("javascript-output");

    const workerDiv = createDiv();
    workerDiv.textContent = `Worker init time: ${workerTime.toFixed(2)} ms`;
    output.appendChild(workerDiv);

    const avgTimeDiv = createDiv();
    avgTimeDiv.textContent = `Avg request time: ${r.avg_time.toFixed(2)} ms`;
    output.appendChild(avgTimeDiv);

    const totalTimeDiv = createDiv();
    totalTimeDiv.textContent = `Total ET: ${r.total_time.toFixed(2)} ms`;
    output.appendChild(totalTimeDiv);

    const requestsDiv = createDiv();
    requestsDiv.textContent = `Requests: ${total_requests}`;
    output.appendChild(requestsDiv);

    const lastValueDiv = createDiv();
    lastValueDiv.textContent = `Last value: ${r.last_value.toFixed(2)}`;
    output.appendChild(lastValueDiv);
}

window.runJsBenchmark = _runJsBenchmark;

File: Astro-TFG/public/scripts/4-manejo-peticiones/prueba-1/version-1/python/main.py
from pyscript import display, document, PyWorker
import js  # type: ignore
import asyncio
import time

worker = None
worker_time = 0.0


async def initialize_worker():
    global worker
    if worker is None:
        base = js.window.location.origin + js.window.document.body.dataset.pyPath
        worker = PyWorker(
            f"{base}worker.py",
            type="pyodide",
            config=f"{base.replace('python', 'json')}pyscript-worker.json"
        )
        await worker.ready


async def launch_worker(event):
    global worker, worker_time
    try:
        t0 = time.perf_counter()
        await initialize_worker()
        worker_time = (time.perf_counter() - t0) * 1000

        num_requests = int(document.getElementById(
            "num-requests-pyscript").value)
        delay = int(document.getElementById("request-delay-pyscript").value)
        base_url = js.location.origin

        res = await worker.sync.do_requests(num_requests, delay, base_url)

        display_result(res, num_requests)

    except Exception as e:
        display(f"Error: {str(e)}", target="pyscript-output")
        print(e)


def display_result(result, num_requests):
    global worker_time
    display(
        f"Worker init time: {worker_time:.2f} ms", target="pyscript-output")
    display(
        f"Avg request time: {result.avg_time:.2f} ms", target="pyscript-output")
    display(
        f"Total ET: {result.total_time:.2f} ms", target="pyscript-output")
    display(
        f"Requests: {num_requests}", target="pyscript-output")
    display(f"Last value {result.last_value:.2f}",
            target="pyscript-output")
    js.window.hideExecutionLoader()


def js_run_py_benchmark(event):
    js.clearCell('pyscript-output')
    asyncio.ensure_future(launch_worker(None))


js.window.run_py_benchmark = js_run_py_benchmark
